### ISSUE TITLE
Two questions about training of Leela Zero

### ISSUE THREAD

**Issue by ionicliquids11** [leela-zero/leela-zero#1981]
Dear Gian-Carlo 

I am a fan of your Leela Zero. Now I am very happy to see that the great progress that you made for Leela Zero and now it is one of the most popular AI for Go players. 

Recently, I noticed that Colaboratory (Google) is giving a cluster of 1,000 Cloud TPUs to researchers for free. Do you think that it is possible that you could modify your Leela zero program which can be run on TPU to help the person like me as well as many other fans who like training Leela Zero.  Sure, I will only try to use TPU in my spare time. It may contribute to the progress of the training of Leela zero.  

Another thing is that when I use GPU of colab in training LZ, sometime it stops due to the Google Colab stops every 12 hours when the game is still runing, do you think it is possible to continue the training from where it stops to save time? 

I am looking forward to your kind reply. 

Best regards

Ionicliquids

A fan of Leela Zero

**Comment by wonderingabout** [comment]
@ionicliquids11 

welcome :) 

i am not gcp but about colab : 
google colab is made for interactive use, it will stop after 1-2 hours if you just passively produce games, as explained in their faq : https://research.google.com/colaboratory/faq.html#gpu-availability

if you want to help, i highly recommend you have a look at google cloud free trial rather : 
https://github.com/gcp/leela-zero#using-a-cloud-provider

**Comment by ionicliquids11** [comment]
Thank you for your suggestion : ) Actually, I am using google cloud free trial for the Leela Zero training. 

Since I noticed that TPU is available in Google Colab and I think that Google Colab is encouraging people to use it.  I just curious if GCP could provide a program for us to use the TPU for training~



> @ionicliquids11
> 
> welcome :)
> 
> i am not gcp but about colab :
> google colab is made for interactive use, it will stop after 1-2 hours if you just passively produce games, as explained in their faq : https://research.google.com/colaboratory/faq.html#gpu-availability
> 
> if you want to help, i highly recommend you have a look at google cloud free trial rather :
> https://github.com/gcp/leela-zero#using-a-cloud-provider

**Comment by gcp** [comment]
Personally I'm not interested in proprietary hardware but if someone would provide patches, sure.

**Comment by gcp** [comment]
> I just curious if GCP could provide a program for us to use the TPU for training

I don't know anything about Google GPUs and I don't have any.

**Comment by ionicliquids11** [comment]
Thanks for your kind reply~

If someone ports a backend that works on TPUs, it could be added to Leela. Personally I'm not very interested in cloud-only single-vendor hardware.

**Comment by ionicliquids11** [comment]
I asked the same question in Minigo's homepage, the following is one answer：

TL;DR It would take a lot of code for moderate gains because of LZ's project setup

I would guess the easiest way to support TPUs would be to extend the IPC (InterProcess Communication) that NTT123 (and later I) wrote. We basically added some code to Network.cpp that pushes the NN inputs into shared memory. We then read them from a python process and evaluated them with theano. At that point you'd read the values in python (similar to this projects old inference_worker.py) and call tensorflow run. This is glossing over a lot of details.

Modify Network to pass weights via grpc or shared memory to python [Easy]
Convert tfprocess.py to tf.estimator [Medium]
Ask GCP to distribute tensorflow weights for write ascii to tf weights converter. [Hard]
Convert inference_worker to use tfprocess.py model and read from shared memory [Easy]
Rewrite some parts of gtp so that all games use the same model.* [Easy]
This would get the code running on TPUs but to see the real speed of TPUs you'd need much larger batches of positions. This means doing virtual loss (This was partially supported in the past I believe)

Figure out a good virtual loss setup [Hard]
* Figure out what to do with 10-20 partial games if the model changes



> > I just curious if GCP could provide a program for us to use the TPU for training
> 
> I don't know anything about Google TPUs and I sure as hell don't have any.
> 
> If someone ports a backend that works on TPUs, it could be added to Leela. Personally I'm not very interested in cloud-only single-vendor hardware.

**Comment by alreadydone** [comment]
I guess we just continue the partial games with the new model; this has to be the case for AZ as the model is trained and updated continuously, no? Batch size <= number of simultaneous games is what ELF did (on V100s), and there seems no reason we can't afford that and has to introduce virtual loss (it seems Mingo does it but I don't know why; virtual loss is definitely in effect with multiple threads during actual play).

FYI I've implemented separation of simulation and children creation/evaluation backup, and am currently tweaking for best performance, in case you really need a large number of positions from one game (this is required for cudnn batching anyway).

**Comment by gcp** [comment]
>FYI I've implemented separation of simulation and children creation/evaluation backup, and am currently tweaking for best performance, in case you really need a large number of positions from one game (this is required for cudnn batching anyway).

Neat!

**Comment by sethtroisi** [comment]
closing old suggestion thread without follow up.
